src/unix/core.c

```diff
@@ -1153,3 +1153,113 @@ int uv_os_tmpdir(char* buffer, size_t* size) {
 
   return 0;
 }
+
+
+void uv_free_passwd(uv_passwd_t* pwd) {
+  if (pwd == NULL)
+    return;
+
+  uv__free(pwd->name);
+  uv__free(pwd->shell);
+  uv__free(pwd->homedir);
+}
+
+
+int uv_get_passwd(uv_passwd_t* pwd) {
+  struct passwd pw;
+  struct passwd* result;
+  char* buf;
+  uid_t uid;
+  size_t bufsize;
+  size_t len;
+  long initsize;
+  int r;
+#if defined(__ANDROID_API__) && __ANDROID_API__ < 21
+  int (*getpwuid_r)(uid_t, struct passwd*, char*, size_t, struct passwd**);
+
+  getpwuid_r = dlsym(RTLD_DEFAULT, "getpwuid_r");
+  if (getpwuid_r == NULL)
+    return -ENOSYS;
+#endif
+
+  if (pwd == NULL)
+    return -EINVAL;
+
+  initsize = sysconf(_SC_GETPW_R_SIZE_MAX);
+
+  if (initsize <= 0)
+    bufsize = 4096;
+  else
+    bufsize = (size_t) initsize;
+
+  uid = geteuid();
+  buf = NULL;
+
+  for (;;) {
+    uv__free(buf);
+    buf = uv__malloc(bufsize);
+
+    if (buf == NULL)
+      return -ENOMEM;
+
+    r = getpwuid_r(uid, &pw, buf, bufsize, &result);
+
+    if (r != ERANGE)
+      break;
+
+    bufsize *= 2;
+  }
+
+  if (r != 0) {
+    uv__free(buf);
+    return -r;
+  }
+
+  if (result == NULL) {
+    uv__free(buf);
+    return -ENOENT;
+  }
+
+  /* Copy the username */
+  len = strlen(pw.pw_name);
+  pwd->name = uv__malloc(len + 1);
+
+  if (pwd->name == NULL) {
+    uv__free(buf);
+    return -ENOMEM;
+  }
+
+  memcpy(pwd->name, pw.pw_name, len + 1);
+
+  /* Copy the shell */
+  len = strlen(pw.pw_shell);
+  pwd->shell = uv__malloc(len + 1);
+
+  if (pwd->shell == NULL) {
+    uv__free(buf);
+    uv_free_passwd(pwd);
+    return -ENOMEM;
+  }
+
+  memcpy(pwd->shell, pw.pw_shell, len + 1);
+
+  /* Copy the home directory */
+  len = strlen(pw.pw_dir);
+  pwd->homedir = uv__malloc(len + 1);
+
+  if (pwd->homedir == NULL) {
+    uv__free(buf);
+    uv_free_passwd(pwd);
+    return -ENOMEM;
+  }
+
+  memcpy(pwd->homedir, pw.pw_dir, len + 1);
+
+  /* Copy the uid and gid */
+  pwd->uid = pw.pw_uid;
+  pwd->gid = pw.pw_gid;
+
+  uv__free(buf);
+
+  return 0;
+}
```

Comment:
It would be neater to allocate the memory for name+shell+homedir with a single malloc call.  That should also make cleanup on ENOMEM easier because you only have to do a single NULL check.

In theory, you don't need an extra allocation because they're all stored in `buf`.  Untested but it should - again, in theory - be possible to release the memory again like this:
```c
void uv_free_passwd(uv_passwd_t* pwd) {
  char* p;
  // ...
  p = pwd->name;
  if (p > pwd->shell) p = pwd->shell;
  if (p > pwd->homedir) p = pwd->homedir;
  free(p);
  // ...
}
```